design-docs/001-virtual-chunk-improvements.md

```diff
@@ -0,0 +1,381 @@
+
+A user can modify the `RepositoryConfig` for their current repo instance, but decide not to save the change to storage. This will be very common as users tweak the configuration during experimentation.
+
+### Setting virtual references
+
+As we mentioned, we want to make as little changes as possible to the way virtual chunks are written to Icechunk by the open source stack. The new `set_virtual_ref` function still takes a `location` url as the main data, adds an optional `checksum` and an optional `validate_containers` argument.
+
+If `validate_containers` is set to `True`, Icechunk will verify in `set_virtual_ref` that there is a `VirtualChunkContainer` with a `prefix` field  that is a prefix of the `location` url passed. An exception is raised if that's not the case and the reference is not stored.
+
+The following examples will work for `set_virtual_ref` called with the given `location` and `validate_containers=True`:
+
+* `location=s3://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=s3`.
+* `location=s3://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=s3://`.
+* `location=s3://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=s3://foo`.
+* `location=s3://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=s3://foo/bar.nc`.
+* `location=tigris://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=tigris`.
+* `location=models://foo/bar.nc` if there is a `VirtualChunkContainer` with `prefix=models`.
+
+The following examples will now work for `set_virtual_ref` called with the given `location` and `validate_containers=True`:
+
+* `location=s3://foo/bar.nc` if there are no `VirtualChunkContainers`
+* `location=gcs://foo/bar.nc` if the only `VirtualChunkContainers` has `prefix=s3`
+
+But calling `set_virtual_ref` in the two last examples will still work with `validate_containers=False`.
+
+The exact algorithm to match a `location` URL to one specific container is explained in a section below.
+
+### New virtual chunk datastructures
+
+We change from
+
+```rust
+pub enum VirtualChunkLocation {
+    Absolute(String),
+    // Relative(prefix_id, String)  // this was a planned approach that remained  unimplemented
+}
+
+pub struct VirtualChunkRef {
+    pub location: VirtualChunkLocation, // this is basically a String for the URL today
+    pub offset: ChunkOffset,
+    pub length: ChunkLength,
+}
+```
+
+To
+
+```rust
+pub struct VirtualChunkLocation(pub String);
+
+type ETag = String;
+
+pub struct SecondsSinceEpoch(pub u32);
+
+pub enum Checksum {
+    LastModified(SecondsSinceEpoch),
+    ETag(ETag),
+}
+
+pub struct VirtualChunkRef {
+    pub location: VirtualChunkLocation,
+    pub offset: ChunkOffset,
+    pub length: ChunkLength,
+    pub checksum: Option<Checksum>,
+}
+```
+
+Using a `u32` for `SecondsSinceEpoch` gives us at least until year 2100.
```

Comment:
I'm imagining a future Y2.1K crisis where we regret this decision and Yottabytes of interplanetary data suddenly become invalid! 😆